pkg/util/sem/sem.go

```diff
@@ -66,26 +69,90 @@ const (
 
 var (
 	semEnabled int32
+	// sysMap record original value of sysvar
+	sysMap      map[string]string
+	mapMutex    sync.Mutex
+	sysMapMutex sync.RWMutex
 )
 
+var isResVarAdmin int32 = 0
+
+// CheckResVarAdmin checks if the user has the RESTRICTED_VARIABLES_ADMIN privilege
+func CheckResVarAdmin(sctx sessionctx.Context) {
+	if sctx == nil {
+		return
+	}
+	checker := privilege.GetPrivilegeManager(sctx)
+	sessionVars := sctx.GetSessionVars()
+	if sessionVars == nil || checker == nil {
+		return
+	}
+	resVarAdmin := checker.RequestDynamicVerification(sessionVars.ActiveRoles, "RESTRICTED_VARIABLES_ADMIN", false)
+	var val int32
+	if resVarAdmin {
+		val = 1
+	} else {
+		val = 0
+	}
+	atomic.StoreInt32(&isResVarAdmin, val)
+}
+
+// IsResVarAdmin returns true if the user has the RESTRICTED_VARIABLES_ADMIN privilege
+func IsResVarAdmin() bool {
+	return atomic.LoadInt32(&isResVarAdmin) == 1
+}
+
+// GetOrigVar Get original system variables
+func GetOrigVar(name string) string {
+	sysMapMutex.RLock()
+	defer sysMapMutex.RUnlock()
+	return sysMap[name]
+}
+
 // Enable enables SEM. This is intended to be used by the test-suite.
 // Dynamic configuration by users may be a security risk.
 func Enable() {
-	atomic.StoreInt32(&semEnabled, 1)
+	if !atomic.CompareAndSwapInt32(&semEnabled, 0, 1) {
+		logutil.BgLogger().Info("SEM enable operation was skipped because it is already enabled")
+		return
+	}
+
+	mapMutex.Lock()
+	defer mapMutex.Unlock()
+	sysMap = make(map[string]string)
 	variable.SetSysVar(variable.TiDBEnableEnhancedSecurity, variable.On)
-	variable.SetSysVar(variable.Hostname, variable.DefHostname)
+
+	cfg := config.GetGlobalConfig()
+	for _, resVar := range cfg.Security.SEM.RestrictedVariables {
+		if resVar.RestrictionType == "replace" {
+			if variable.IsVarExists(resVar.Name) {
+				originalValue := variable.GetSysVar(resVar.Name).Value
+				if _, ok := sysMap[resVar.Name]; !ok {
+					sysMap[resVar.Name] = originalValue
+					variable.SetSysVar(resVar.Name, resVar.Value)
```

Comment:
Will this conflict with the global variable loaded later?

Reply:
> Will this conflict with the global variable loaded later?
Won't. This only modifies the variables when SEM starts. If system variables are set when TiDB is subsequently run, the current modifications will be overwritten.